### PR TITLE
Add View Hierarchy

### PR DESCRIPTION
**Pull request by krystofwoldrich**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description
Works both on iOS and Android but it can't be merged until stable releases of the native SDKs.

## :bulb: Motivation and Context
- https://github.com/getsentry/team-mobile/issues/64

## :green_heart: How did you test it?
sample app, 

## :pencil: Checklist

- [x] I reviewed submitted code
- [ ] I added tests to verify changes
- [x] All tests passing
- [ ] No breaking changes

## :crystal_ball: Next steps
- [x] Depends on [sentry-cocoa v8 release](https://github.com/getsentry/sentry-cocoa/tree/8.0.0)
- Will be part of v5 RN SDK
- [x] Depends on https://github.com/getsentry/sentry-java/pull/2440
- [x] Fix error logs spam when capturing the hierarchy in android https://github.com/getsentry/sentry-java/pull/2440#issuecomment-1378645411

### CODE REVIEW

**Comment by marandaneto** [comment]
@krystofwoldrich CI is failing (E2E), but not sure if its because of that.
Is this ready to be reviewed?

**Comment by marandaneto** [comment]
We'd need to enable this feature on the docs as well, See [Flutter docs](https://docs.sentry.io/platforms/flutter/enriching-events/viewhierarchy/).

**Comment by marandaneto** [comment]
@krystofwoldrich why do we base on top of `cocoa-v8`? shouldn't be in the RN 5 branch?

**Comment by marandaneto** [comment]
> @krystofwoldrich why do we base on top of `cocoa-v8`? shouldn't be in the RN 5 branch?

Ah ok because https://github.com/getsentry/sentry-react-native/pull/2707 isn't merged yet.

**Comment by krystofwoldrich** [comment]
@marandaneto CI was failing because of https://github.com/getsentry/sentry-react-native/pull/2770

I've merged into v5 and cocoa-v8 branches so it should be good now.

File: src/js/integrations/viewhierarchy.ts
Comment (marandaneto) [inline src/js/integrations/viewhierarchy.ts:39]: H: `attachmentType` should be `event.view_hierarchy`

File: android/src/main/java/io/sentry/react/RNSentryModuleImpl.java
Comment (marandaneto) [inline android/src/main/java/io/sentry/react/RNSentryModuleImpl.java:410]: M: This code already lives in the `ViewHierarchyEventProcessor` Android, maybe the processor exposes this function as static and RN can reuse it, right now it's duplicating code which increases bundle size.

File: android/src/main/java/io/sentry/react/RNSentryModuleImpl.java
Comment (marandaneto) [inline android/src/main/java/io/sentry/react/RNSentryModuleImpl.java:395]: H: I'd instead prefer that `ViewHierarchyEventProcessor` exposes a function that returns the array of bytes directly since the Android SDK already knows everything needed, this function only needs the `Activity`, less code to maintain, less duplication, and less memory usage since you don't need to create your own serializer, iOS is already like that.

File: src/js/integrations/viewhierarchy.ts
Comment (krystofwoldrich) [reply src/js/integrations/viewhierarchy.ts:39]: Thanks.

File: android/src/main/java/io/sentry/react/RNSentryModuleImpl.java
Comment (krystofwoldrich) [reply android/src/main/java/io/sentry/react/RNSentryModuleImpl.java:410]: I've updated the view hierarchy event processor so all the checks are done in it.
https://github.com/getsentry/sentry-java/pull/2492

File: android/src/main/java/io/sentry/react/RNSentryModuleImpl.java
Comment (marandaneto) [reply android/src/main/java/io/sentry/react/RNSentryModuleImpl.java:395]: @krystofwoldrich the change has to live in the Android SDK, so @markushi and @romtsn maybe can assist you with the changes or do it for you.

**Comment by krystofwoldrich** [comment]
I've simplified the view hierarchy function based on https://github.com/getsentry/sentry-java/pull/2492

**Comment by marandaneto** [comment]
> I've simplified the view hierarchy function based on
> 
> * [chore(internal): Refactor view hierarchy to be more hybrid SDK friendly sentry-java#2492](https://github.com/getsentry/sentry-java/pull/2492)

Let me know once this is ready to be reviewed again since there are conflicts to be solved yet + using the new Android API.

**Comment by krystofwoldrich** [comment]
@marandaneto `sentry-java` is merged. This is ready for the final review.

File: android/src/main/java/io/sentry/react/RNSentryModuleImpl.java
Comment (marandaneto) [inline android/src/main/java/io/sentry/react/RNSentryModuleImpl.java:200]: H: This would not work if [manual init](https://docs.sentry.io/platforms/react-native/manual-setup/native-init/) is not enabled because we won't have the options ref.
An option is to read the serializer from the HubAdapter `HubAdapter.getInstance().getOptions().getSerializer()`.

File: android/src/main/java/io/sentry/react/RNSentryModuleImpl.java
Comment (marandaneto) [inline android/src/main/java/io/sentry/react/RNSentryModuleImpl.java:403]: M: What if `ViewHierarchyEventProcessor` already has a method called `snapshotViewHierarchyAsData` or something that returns the `bytes` directly as suggested before, right now the Hybrid SDK needs to take care of serialization and this is the Native SDK's responsibility IMO.
I'd not block this PR because of that, but that was the suggestion before the refactoring, wondering why we've not taken this approach.

File: src/js/integrations/viewhierarchy.ts
Comment (marandaneto) [reply src/js/integrations/viewhierarchy.ts:39]: This comment was resolved but I didn't see any change related to `attachmentType`.

File: android/src/main/java/io/sentry/react/RNSentryModuleImpl.java
Comment (krystofwoldrich) [reply android/src/main/java/io/sentry/react/RNSentryModuleImpl.java:403]: That related [to the first comment](https://github.com/getsentry/sentry-react-native/pull/2708/files/f64ac6fa97a762dba4cb8b21601e73c76154b96e#r1092125493) because I didn't know how to get a reference of the serializer in the ViewHierarchyEventProcessor.

Comment (krystofwoldrich) [reply android/src/main/java/io/sentry/react/RNSentryModuleImpl.java:395]: I've opened PR in java https://github.com/getsentry/sentry-java/pull/2508

But as you said we can finish this and when the java PR is released we can simplify it in a new PR?

File: sample/src/App.tsx
Comment (marandaneto) [inline sample/src/App.tsx:31]: Are you sure you wanna add the hint back? it caused problems already.

File: android/src/main/java/io/sentry/react/RNSentryModuleImpl.java
Comment (marandaneto) [reply android/src/main/java/io/sentry/react/RNSentryModuleImpl.java:395]: Fine for me, just address the other comments.

File: src/js/integrations/viewhierarchy.ts
Comment (krystofwoldrich) [reply src/js/integrations/viewhierarchy.ts:40]: I remember fixing it... strange. Now it's here.

File: sample/src/App.tsx
Comment (krystofwoldrich) [reply sample/src/App.tsx]: I remember removing it. :( Now it's gone.

**Comment by krystofwoldrich** [comment]
One last thing
- https://github.com/getsentry/sentry-java/pull/2513

**Review by marandaneto — APPROVED** [review]
LGTM after upgrading the Android SDK with the latest changes (dupe of VH attachments)

**Comment by krystofwoldrich** [comment]
Waiting for https://github.com/getsentry/sentry-java release, likely `6.13.1`.

**Pull request merged**
